Support extra output field in TOP function

przemekwitek opened the pull request:
There are cases where we would like to answer queries like:
"give me list of employees sorted by their salaries"
"give me list of countries sorted by their area"
etc.
Currently, the `TOP` aggregation function only supports one field which is used as both sort field and the output field.
This PR enhances the `TOP` function by adding an optional parameter `mapToField` which is the field we want as the output. The sorting will still be performed on the `field`, just like today.

Example:
"give me salaries of 3 youngest employees by gender"
```
FROM employees
| STATS youngest_employees = TOP(birth_date, 3, "desc"),
        youngest_employees_salaries = TOP(birth_date, 3, "desc", salary)
  BY    gender
| SORT gender
| KEEP gender, youngest_employees, youngest_employees_salaries
;

gender:keyword | youngest_employees:datetime                                                    | youngest_employees_salaries:integer
F              | [1964-10-18T00:00:00.000Z, 1964-06-02T00:00:00.000Z, 1963-03-21T00:00:00.000Z] | [25976, 56371, 43602]
M              | [1965-01-03T00:00:00.000Z, 1964-06-11T00:00:00.000Z, 1964-04-18T00:00:00.000Z] | [37702, 45656, 46595]
null           | [1963-06-07T00:00:00.000Z, 1963-06-01T00:00:00.000Z, 1961-05-02T00:00:00.000Z] | [48735, 45797, 61358]
;
```

Fixes https://github.com/elastic/elasticsearch/issues/128630

Commit:
Support extra field in TOP function

Commit:
Update docs/changelog/135434.yaml

elasticsearchmachine commented:
Hi @przemekwitek, I've created a changelog YAML for you.

ivancea reviewed (commented):
Amazing work!
I'm leaving the TopTests part, as I understood you're still working on that. I'll complete the review after it. Everything else was reviewed, and looks good

Commit:
Apply review comments

przemekwitek commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/stats_top_map.csv-spec:
FYI: In order to support the TOP(a, a) case, I added a change in `Top.surrogate` method. PTAL

przemekwitek commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/sort/X-BucketedSort.java.st:
```diff
@@ -237,8 +318,15 @@ $endif$
-        return getOrder().reverseMul() * $Wrapper$.compare(lhs, rhs) < 0;
+    private boolean betterThan($type$ lhs, $type$ rhs$if(hasExtra)$, $Extratype$ lhsExtra, $Extratype$ rhsExtra$endif$) {
+        int res = $Wrapper$.compare(lhs, rhs);
+$if(hasExtra)$
```
Ok, now I see why this consistent order mattered to me. Some of the BWC tests are failing due to the order of entries being different than expected. 
See for example https://gradle-enterprise.elastic.co/s/evcqwpf64bm3m/tests/task/:x-pack:plugin:esql:qa:server:multi-clusters:v9.0.9%23bwcTest/details/org.elasticsearch.xpack.esql.ccq.MultiClusterSpecIT/test%20%7Bcsv-spec:stats_top.EmpNoByHireDateByBirthYear%7D?top-execution=1 .
So we have 2 options: restore the consistent order or relax the tests somehow so that the order does not matter, e.g.: with `MV_SORT`. WDYT?

Commit:
- Add enhanced TOP tests to TopTests.java
- Fix grow() method in X-BucketedSort
- Add `date` type to supported types of `outputField`

Commit:
Fix style with spotless

Commit:
Fix TOP's type resolution and add appropriate tests

Commit:
Regenerate TopXYAggregatorFunction classes.

Commit:
Add tests for new surrogates

Commit:
Apply spotless

Commit:
Fix stats_top.empNoByHireDateByBirthYear by fixing the order of rows

Commit:
Apply review comment

Commit:
[CI] Auto commit changes from spotless

Commit:
Add "begin/end generated imports" directives to X-BucketedSort.java.st file

Commit:
Make spotless happy

ivancea commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/aggregate/TopTests.java:
```diff
@@ -285,32 +450,75 @@ public static Iterable<Object[]> parameters() {
-                .toList();
+            TestCaseSupplier.TypedData outputFieldTypedData;
+            List<?> expected;
+            if (outputFieldSupplied) {
```
nit: To calculate the expected, maybe I would just always treat it as if the `outputParam == fieldParam` when `outputFieldSupplied == false`, and remove the old branch.

Take this with a grain of salt; I just feel like this method is getting quite chonky, but not a big deal as it's well localized 🚀

przemekwitek replied:
> nit: To calculate the expected, maybe I would just always treat it as if the outputParam == fieldParam when outputFieldSupplied == false, and remove the old branch.

Done.

ivancea reviewed (commented):
Looks good, thanks!

Added two documentation comments that I think are important though 👀

przemekwitek commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/stats_top.csv-spec:
```diff
@@ -312,3 +312,96 @@ top_salaries:integer
 ;
 
+
+youngestEmployees
```
> nit: We usually name test cases based on the technical aspect we're testing, instead of the data we're getting. It helps better understand which edge cases we're testing without having to read the query (usually).

Done.

Commit:
Apply review comments

Commit:
spotless

Commit:
Include description in appliesTo

Commit:
Add generated appliesTo description

Commit:
Revert surrogate unit tests

Commit:
Adapt Top.java after merge

przemekwitek commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/Top.java:
I added `description` to docs generation. Let's see how it looks like.

Commit:
Add missing multi-match description

leemthompo commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/Top.java:
```diff
@@ -66,7 +88,13 @@ public class Top extends AggregateFunction
+        examples = @Example(file = "stats_top", tag = "top"),
+        appliesTo = {
+            @FunctionAppliesTo(
+                lifeCycle = FunctionAppliesToLifecycle.COMING,
```
Should use PREVIEW or GA here instead of COMING

leemthompo reviewed (commented):
Couple things about applies_to

leemthompo commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/Top.java:
```diff
@@ -66,7 +88,13 @@ public class Top extends AggregateFunction
+            @FunctionAppliesTo(
+                lifeCycle = FunctionAppliesToLifecycle.COMING,
+                version = "9.3.0",
+                description = "Support for optional order and outputField parameters is only available from 9.3.0"
```
You can use [inline applies_to](https://github.com/elastic/elasticsearch/blob/main/docs/reference/query-languages/esql/README.md#inline-applies_to-metadata) in the outputField param description

````
description = "The extra field that, if present, will be the output of the TOP call instead of `field`.{applies_to}`stack: ga 9.3.`"
````

Just sharing Slack convo for anybody following along :)

przemekwitek replied:
Done.

ivancea replied:
Question: I see this is rendered as:
<img width="697" height="95" alt="image" src="https://github.com/user-attachments/assets/501d0eb7-8701-4935-9f02-9deef1e4ae81" />
Which doesn't show anything like "coming in a future release" nor the version it will be on. Is that right? How do users know that the function isn't available in 9.1, for example?

Commit:
Revert FunctionAppliesTo changes.
Put inline applies_to in the outputField param section.

Commit:
Fix style

Commit:
Enhance description

przemekwitek commented on docs/changelog/135434.yaml:
```diff
@@ -0,0 +1,6 @@
+pr: 135434
+summary: Support extra field in TOP function
```
Done.

przemekwitek commented:
@ivancea , I've addressed the comments. The tests are passing and the docs are generated without issues.
PTAL

ivancea reviewed (approved):
Looks good!
Added this question [here](https://github.com/elastic/elasticsearch/pull/135434#discussion_r2502789584), as the final point on applies_to. I would wait until we're 100% sure about that; but otherwise, go for it :shipit:

florent-leborgne commented on docs/reference/query-languages/esql/_snippets/functions/parameters/top.md:
```diff
@@ -11,3 +11,6 @@
 :   The order to calculate the top values. Either `asc` or `desc`, and defaults to `asc` if omitted.
 
+`outputField`
+:   The extra field that, if present, will be the output of the TOP call instead of `field`.{applies_to}`stack: ga 9.3.`
```
```suggestion
:   The extra field that, if present, will be the output of the TOP call instead of `field`.{applies_to}`stack: ga 9.3`
```
The extra `.` was making this an incorrect value

przemekwitek replied:
Done.

Commit:
Fix the docs typo.

Commit:
Fix typo in proper file.
